pyo3-macros-backend/src/pymethod.rs

```diff
@@ -439,6 +439,19 @@ fn impl_traverse_slot(
             i.e. `Python::with_gil` will panic."));
     }
 
+    if let FnType::Fn(SelfType::TryFromBoundRef(span))
+    | FnType::Fn(SelfType::Receiver {
+        mutable: true,
+        span,
+    }) = spec.tp
+    {
+        bail_spanned! { span =>
+            "__traverse__ may not take a receiver other than `&self`. Usually, an implementation of `__traverse__` \
+            should do nothing but calls to `visit.call`. Most importantly, safe access to the GIL is prohibited \
```

Comment:
Not really new because the error message above has the same issue but since we are here: The reference to `visit.call` could be somewhat opaque if the user has not named the argument `visit: PyVisit`. This would trigger a different error message itself, but I think we might want to introduce the `visit` identifier somehow, e.g.

```
an implementation of `__traverse__(&self, visit: PyVisit)`
```

or something like that.

Really not sure about this particular proposal, but the `visit` identifier should be bound somehow.

Reply:
Sounds reasonable, I added the signature hint to both error messages, and an improved error, if the signature is wrong.
The error message is still not so nice if the return type is wrong, but its workable I think.